### DISCUSSION TITLE
Trealla briefly mentioned in Prolog: The Next 50 Years book

### DISCUSSION THREAD

**Discussion by guregu** [trealla-prolog/trealla-prolog#228] [General]
Trealla was mentioned in the paper [The Janus System: A Bridge to New Prolog Applications](https://link.springer.com/chapter/10.1007/978-3-031-35254-6_8) which includes a survey of Prolog implementations and their embedding interfaces. Too bad they missed our WASM, JS, and Go ports :)

<img width="553" alt="Screenshot 2023-07-01 at 3 17 08" src="https://github.com/trealla-prolog/trealla/assets/131059/13dde9cc-85e3-4662-877d-7dd8a42eb877">

The FFI interface described in the paper sounds pretty cool. It can automatically represent Python objects in Prolog and vice versa. My ports basically require you to work with Prolog terms for everything, and they communicate with JSON so they are definitely slower than shuffling raw memory around like in the paper. The SWI WASM port also has some cool syntax for calling JS code from Prolog that I need to take a closer look at.

Anyway just wanted to point this out. Haven't read all of the papers yet but there's also some stuff about WASM-based Prolog running in browsers for teaching that caught my eye. I think trealla-js is already very useable, and I think it could prove useful for those kinds of situations as well. I will be playing around with more embedding stuff for sure.

Thanks for Trealla, @infradig!

**Comment by infradig** [comment]
I'm familiar with the XSB/Python interface and have worked with Theresa in
implementing it elsewhere. While adding it to Trealla would be
straightforward it's not something I feel the need to do.

Thanks for your support.
